### PR TITLE
Split CSIDriverController into two control loops

### PR DESCRIPTION
**Pull request by bertinatto**
This was recommended by @deads2k in a code review: https://github.com/openshift/library-go/pull/829#discussion_r455884233

CSIDriverControllerServiceController creates the CSI Controller Service deployment. 
CSIDriverNodeServiceController creates the CSI Node Service DaemonSet.

### CODE REVIEW

**Comment by bertinatto** [comment]
/assign @sttts 
Stefan, could you take a look at this?

**Comment by bertinatto** [comment]
CC @openshift/storage

File: pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	}
+
+	if ok, msg := isProgressing(opStatus, deployment); ok {
+		progressingCondition.Status = opv1.ConditionTrue
+		progressingCondition.Message = msg
+		progressingCondition.Reason = "AsExpected"
+	}
+
+	updateStatusFn := func(newStatus *opv1.OperatorStatus) error {
+		// TODO: set ObservedGeneration (the last stable generation change we dealt with)
```
Comment (jsafrane) [inline pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go:159]: Setting ObservedGeneration is IMO hard, as there are two independent controller that need to agree on its value.

Comment (deads2k) [reply pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go:159]: > Setting ObservedGeneration is IMO hard, as there are two independent controller that need to agree on its value.

This doesn't have a good solution in any controller yet.  Generically across other operators many different pieces need to agree.  I don't think this is significantly worse than other solutions we've attempted.

**Comment by deads2k** [comment]
I like this a lot.  @p0lyn0mial @marun I think this is a good point to start from for creating generic management of deployments and daemonsets. 

/approve

I'll leave lgtm with @jsafrane 
/assign @jsafrane

File: pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	driverImageEnvName              = "DRIVER_IMAGE"
```
[the remark this answers is not in the document]
Comment (bertinatto) [reply pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go:25]: Technically they don't refer to the same manifest file, so if I change this const it'll look for the placeholder in the controller.yaml only. I can create a new package if you prefer

File: pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+	// Log level
+	logLevel := getLogLevel(spec.LogLevel)
+	pairs = append(pairs, []string{"${LOG_LEVEL}", strconv.Itoa(logLevel)}...)
+
+	replaced := strings.NewReplacer(pairs...).Replace(string(manifest))
+	return []byte(replaced)
+}
+
+func getLogLevel(logLevel opv1.LogLevel) int {
```
[the remark this answers is not in the document]
Comment (bertinatto) [reply pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go:206]: I could create a new package, but I found it overkill. Personally I prefer it this way, but I can create a new package if you insist

File: pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+//
+// 2. Log level
+//
+// The controller can also override the log level passed in to the CSI driver container.
+// In order to do that, the placeholder ${LOG_LEVEL} from the manifest file is replaced with the value specified
+// in the OperatorClient resource (Spec.LogLevel).
+//
+// This controller produces the following conditions:
+//
+// <name>Available: indicates that the CSI Controller Service was successfully deployed.
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go:62]: It waits for `deployment.Status.AvailableReplicas > 0`

File: pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go:206]: Hm, the code is copied all over the place (in every operator?). Ok, let's leave it as it is for now.

File: pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bertinatto) [reply pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go]: Understood. Fixed

File: pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bertinatto) [reply pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go]: Actually there's a package that already exposes this function: https://github.com/bertinatto/library-go/blob/35c2f0bd29166ad2b495a9aec3df51b71b1dc32c/pkg/operator/loglevel/util.go#L13-L26

The only difference to our function is that TraceAll is 8, not 100. That should be OK, though, so I'm using that one instead.

**Pull request merged**